### PR TITLE
Allow index for ObjectStorage to be int

### LINKED ISSUE #47: Parameter expects string, int<0, max> given.
Error message:

Parameter 1 $value of method TYPO3\CMS\Extbase\Persistence\ObjectStorage<Vendor\Ext\Domain\Model\Day>::offsetGet() expects string, int<0, max> given. 

Sample Code:
```

/** @var TYPO3\CMS\Extbase\Persistence\ObjectStorage<Vendor\Ext\Domain\Model\Day> $days */
$days = getDays();

$day = $days->offsetGet($days->count() - 1);


```

I believe the problem is here:
vendor/saschaegerer/phpstan-typo3/stubs/ObjectStorage.stub
```

/**
 * @template T
 * @implements \ArrayAccess<string, T>
 * @implements \Iterator<string, T>
 */
class ObjectStorage implements \ArrayAccess, \Iterator
{
}

```

But it should be 

/**
 * @template T
 * @implements \ArrayAccess<int|string, T>
 * @implements \Iterator<int|string, T>
 */
class ObjectStorage implements \ArrayAccess, \Iterator
{
}

```

### PR DESCRIPTION
**Pull request by tlueder**
Fixes #47

### CODE REVIEW

**Comment by sascha-egerer** [comment]
@tlueder Thank you for your contribution! Would it be possible for you to add a test? If you don't know how just tell me. Then i'll take care.

**Comment by tlueder** [comment]
I'm looking at it right now. And have no clue how to do that in this setting.

**Comment by sascha-egerer** [comment]
@tlueder Just had a look at it and noticed hat `string` should not be possible at all. There is also some other special behavior in the ObjectStorage that must be implemented by a reflection like the behavior of "offsetGet".

I'll prepare a new patch for that, soon.

**Comment by tlueder** [comment]
@sascha-egerer Thanks.

**Comment by sascha-egerer** [comment]
Oh i totally forgot about this one... sorry. Will try to have a look soon.

**Comment by sabbelasichon** [comment]
@sascha-egerer Looks good to me. In the original ObjectStorage class the methods like offsetGet have a docblock with int and TEntity specified. So i think that the string should be maybe replaced by an integer anyways for the ArrayAccess.
Have a look at: https://github.com/TYPO3/typo3/blob/main/typo3/sysext/extbase/Classes/Persistence/ObjectStorage.php#L215

**Comment by sascha-egerer** [comment]
> @sascha-egerer Looks good to me. In the original ObjectStorage class the methods like offsetGet have a docblock with int and TEntity specified. So i think that the string should be replaced by an integer anyways for the ArrayAccess. Have a look at: https://github.com/TYPO3/typo3/blob/main/typo3/sysext/extbase/Classes/Persistence/ObjectStorage.php#L215

But this is not the case for TYPO3 10.4 https://github.com/TYPO3/typo3/blob/10.4/typo3/sysext/extbase/Classes/Persistence/ObjectStorage.php#L201

An then there is this behavior that returns the object itself or the information of the object depending on the param type. This is what i meant with special behavior. But there is now "conditinal return types" since PHPStan 1.6 so this can be used to write the "condition" https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types#conditional-return-types

**Comment by sabbelasichon** [comment]
@sascha-egerer Thanks. But i would guess that changing string to an int in the stub will fix the problem. The stub has precedence over the original class. You would not pass a string to the offsetGet. But maybe i don´t get it completely.

**Comment by sascha-egerer** [comment]
> @sascha-egerer Thanks. But i would guess that changing string to an int in the stub will fix the problem. The stub has precedence over the original class. You would not pass a string to the offsetGet, would you? But maybe i don´t get it completely.

Ok, as discussed in Slack we should do it like that. We'll change the type to be only int as this is what is allowed since TYPO3 11. Strings should not be used anymore.

The other change with the conditional return type will be implemented in a separate PR.

File: stubs/ObjectStorage.stub [hunk 1]
```diff
@@ -3,8 +3,8 @@ namespace TYPO3\CMS\Extbase\Persistence;
 
 /**
  * @template T
- * @implements \ArrayAccess<string, T>
- * @implements \Iterator<string, T>
+ * @implements \ArrayAccess<int|string, T>
+ * @implements \Iterator<int|string, T>
```
Comment (sascha-egerer) [inline stubs/ObjectStorage.stub:7]: `ObjectStorage::key()`, that is the implementation for `\Iterator::key()` does return a string.
So `int` is not correct here as `Iterator` will use `int` now with this annotation. So we can't do it like that.
Internally this iterator does also use a string and not an int. So i would suggest to keep this with string and just change the annotations of the methods that have a deviating behavior

**Review by sascha-egerer — CHANGES_REQUESTED** [review]

**Comment by sascha-egerer** [comment]
> > @sascha-egerer Thanks. But i would guess that changing string to an int in the stub will fix the problem. The stub has precedence over the original class. You would not pass a string to the offsetGet, would you? But maybe i don´t get it completely.
> 
> Ok, as discussed in Slack we should do it like that. We'll change the type to be only int as this is what is allowed since TYPO3 11. Strings should not be used anymore.
> 
> The other change with the conditional return type will be implemented in a separate PR.
> 
> But we can still add a test for this.

Just had another look at it. I'll push a new version

**Comment by sascha-egerer** [comment]
Don't know why build fails. This PR does now depend on #56 #57 and #58. So if these are merged we should check this one again.

**Review by sascha-egerer — APPROVED** [review]

**Comment by sascha-egerer** [comment]
@sabbelasichon Could you please have a look if this looks good to you? I'll merge it if it's fine for you

**Comment by sabbelasichon** [comment]
Looks good. Was testing it in the meantime. Thanks for merging.

**Pull request merged**